src/guacscanner/guacscanner.py

```diff
@@ -0,0 +1,811 @@
+).format(
+    table=psycopg.sql.Identifier("guacamole_connection_permission"),
+    entity_id_field=psycopg.sql.Identifier("entity_id"),
+    connection_id_field=psycopg.sql.Identifier("connection_id"),
+    permission_field=psycopg.sql.Identifier("permission"),
+)
+DELETE_CONNECTION_PERMISSIONS_QUERY = psycopg.sql.SQL(
+    """DELETE FROM {table} WHERE {connection_id_field} = %s;"""
+).format(
+    table=psycopg.sql.Identifier("guacamole_connection_permission"),
+    connection_id_field=psycopg.sql.Identifier("connection_id"),
+)
+
+
+def entity_exists(db_connection, entity_name, entity_type):
+    """Return a boolean indicating whether an entity with the specified name and type exists."""
+    with db_connection.cursor() as cursor:
+        logging.debug(
+            "Checking to see if an entity named %s of type %s exists in the database.",
+            entity_name,
+            entity_type,
+        )
+        cursor.execute(ENTITY_COUNT_QUERY, (entity_name, entity_type))
+        count = cursor.fetchone()["count"]
+        if count != 0:
+            logging.debug(
+                "An entity named %s of type %s exists in the database.",
+                entity_name,
+                entity_type,
+            )
+        else:
+            logging.debug(
+                "No entity named %s of type %s exists in the database.",
+                entity_name,
+                entity_type,
+            )
+
+        return count != 0
+
+
+def get_entity_id(db_connection, entity_name, entity_type):
+    """Return the ID corresponding to the entity with the specified name and type."""
+    logging.debug("Looking for entity ID for %s of type %s.", entity_name, entity_type)
+    with db_connection.cursor() as cursor:
+        logging.debug(
+            "Checking to see if any entity named %s of type %s exists in the database.",
+            entity_name,
+            entity_type,
+        )
+        cursor.execute(ENTITY_ID_QUERY, (entity_name, entity_type))
+
+        # Note that we are assuming there is only a single match.
+        return cursor.fetchone()["entity_id"]
+
+
+def add_user(
+    db_connection: psycopg.Connection,
+    username: str,
+    password: str = None,
+    salt: bytes = None,
+) -> int:
+    """Add a user, returning its corresponding entity ID.
+
+    If password (salt) is None (the default) then a random password
+    (salt) will be generated for the user.
+
+    Note that the salt should be an array of bytes, while the password
+    should be an ASCII string.
+
+    """
+    logging.debug("Adding user entry.")
+
+    if password is None:
+        # Generate a random password consisting of ASCII letters and
+        # digits
+        alphabet = string.ascii_letters + string.digits
+        password = "".join(
+            secrets.choice(alphabet) for i in range(DEFAULT_PASSWORD_LENGTH)
+        )
+    if salt is None:
+        # Generate a random byte array
+        salt = secrets.token_bytes(DEFAULT_SALT_LENGTH)
```

Comment:
Related to suggestion above:
```suggestion
        salt = secrets.token_bytes(DEFAULT_PASSWORD_SALT_LENGTH)
```